src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -231,4 +224,79 @@ protected void performCleanup() {
             // no action
         }
     }
+
+    /**
+     * A specialized implementation that tracks phantom cleanables.
+     * Backing storage is expanded and trimmed automatically.
+     * Insert/remove run in amortized constant time.
+     */
+    static final class PhantomCleanableList {
+        private final ArrayList<PhantomCleanable<?>> list = new ArrayList<>();
+        private int maxIdx;
+
+        /**
+         * Returns true if cleanable list is empty.
+         *
+         * @return true if the list is empty
+         */
+        public boolean isEmpty() {
+            synchronized (list) {
+                return list.isEmpty();
+            }
+        }
+
+        /**
+         * Insert this PhantomCleanable in the list.
+         */
+        public void insert(PhantomCleanable<?> phc) {
+            synchronized (list) {
+                // Inserting at the end, record the indexes.
+                int size = list.size();
+                phc.index = size;
+                maxIdx = size;
+                list.add(phc);
+            }
+        }
+
+        /**
+         * Remove this PhantomCleanable from the list.
+         *
+         * @return true if Cleanable was removed or false if not because
+         * it had already been removed before
+         */
+        public boolean remove(PhantomCleanable<?> phc) {
+            synchronized (list) {
+                int thisIdx = phc.index;
+                if (thisIdx == -1) {
+                    // Was already removed from the list.
+                    return false;
+                }
+
+                // Unlink PhantomCleanable.
+                phc.index = -1;
+
+                int lastIdx = list.size() - 1;
+                if (lastIdx != thisIdx) {
+                    // Move the last element at current index, overwriting it.
+                    // Update its index to a new location.
+                    PhantomCleanable<?> last = list.get(lastIdx);
+                    last.index = thisIdx;
+                    list.set(thisIdx, last);
+                }
+
+                // Cut the tail. Runs in constant time.
+                list.remove(lastIdx);
+
+                // Capacity control: trim the backing storage if it looks like
+                // we have a lot of wasted space there.
+                if (list.size() > maxIdx * 2) {
+                    list.trimToSize();
+                    maxIdx = list.size() - 1;
+                }
+
+                // Success!
+                return true;
+            }
+        }
```

Comment:
It looks like the new `PhantomCleanableList.remove()` puts 100% trust in the correctness of the PhantomCleanable's `index`.

What would you think of adding some sort of check/assert that `list.get(phc.index) == phc` ?

Reply:
Right, tracking indexes is the part of the algo. I added asserts in most important sites, `java/lang` still passes with `-esa`. Please take a look again.

Reply:
Better - thanks!